pkg/util/gate/gate.go

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/thanos-io/thanos/blob/main/pkg/gate/gate.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Thanos Authors.
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/util/gate/gate.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors.
+
+package gate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var ErrMaxConcurrent = errors.New("max concurrent requests inflight")
```

Comment:
Would be very nice to make it a global error (with runbook). Since this is experimental, I wouldn't block on it, but could you open an issue about it, please?